### PR TITLE
support model blind index

### LINKED ISSUE #4: support model blind index
## Problem Statement

currently, after the developer integrated this package on an Eloquent model, there wouldn’t be able to search the value of the encrypted field anymore. however, it might be the common use case that developers need to search certain fields even if that one has been encrypted.

## Requirements

1. should make the encrypted field searchable
2. should provide methods on query builder to search encrypted field

### PR DESCRIPTION
**Pull request by ericHao22**
### Blind index planning

to achieve searching on encrypted fields, we use a strategy called blind indexing. its idea is to store a hash value of the plaintext in a separate column and would it will be used for searching.

- Plan A: Create a polymorphism table for storing all blind indexes
  - Pros
    - developer wouldn’t need to create migration to add blind index column for every table
    - package wouldn’t need to handle customization（ex. blind index column name）
  - Cons
    - encryption on multiple tables would create a huge amount of data in the blind index table, and that might cause potential performance issue
    - **developer should need make sure the updating operation across the model table and index table is at the same transaction to avoid race condition**
- Plan B: Create an additional column on the original table for storing the blind index
  - Pros
    - implementation is simpler for unique index
    - searching and updating on the same table/B-tree would make performance better
    - can make sure an updating operation of raw value and hash value is consistent & atomic
  - Cons
    - need to create many additional columns for each searchable column. (email\_bidx, phone\_bidx, address_bidx, etc...)

we choose to use Plan B because we should highly consider performance issues when developing this package.

### Backend

1. extract `Encrypter` class to encapsulate logic related with data row encryption
2. make `Encrypter` support generating blind indices of data row
3. make `KeyManager` support retrieving/generating hash key
4. make `Securable` support updating blind indices when model saved
   1. should use same hash key to generate blind indices
5. make `Securable` extend query builder to search encrypted value
   1. `where`
   2. `orWhere`

### CODE REVIEW

File: tests/Migrations/2023_03_02_110100_add_email_bidx_column_to_users_table.php [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+{
+    /**
+     * Run the migrations.
+     *
+     * @return void
+     */
+    public function up()
+    {
+        Schema::table('users', function (Blueprint $table) {
+            $table->text('email_bidx')->nullable(true);
```
Comment (FishChen1202) [inline tests/Migrations/2023_03_02_110100_add_email_bidx_column_to_users_table.php:17]: 以後如果有新的欄位要加密 也都是需要自己 create migration file 嗎 還是這個套件有提供指令來做掉這一塊？

**Review by FishChen1202 — APPROVED** [review]
r+, with a comment. thanks!

File: tests/Migrations/2023_03_02_110100_add_email_bidx_column_to_users_table.php (hunk 1 above)
Comment (ericHao22) [reply tests/Migrations/2023_03_02_110100_add_email_bidx_column_to_users_table.php:17]: 目前需要使用者自己建立，未來有考慮提供指令來生成，但想像的是可能一次只能針對一個欄位產生migration file

**Review by koshuang — APPROVED** [review]
r+, thanks!

**Pull request merged**